data/blog/2025/2025-03-18-mezzio101-what-defines-a-middleware-architecture.md

```diff
@@ -0,0 +1,123 @@
+---
+id: 2025-03-18-mezzio101-what-defines-a-middleware-architecture
+author: bidi
+title: 'Mezzio101: What Defines a Middleware Architecture?'
+draft: false
+public: true
+created: '2025-03-18T11:00:00-01:00'
+updated: '2025-03-18T11:00:00-01:00'
+tags:
+  - middleware
+  - software architecture
+  - framework
+openGraphImage: '2025-03-18-mezzio101-what-defines-a-middleware-architecture.png'
+openGraphDescription: 'Mezzio101: What Defines a Middleware Architecture?'
+---
+
+Nowadays, applications are getting more and more complex.
+The classic way of executing sequential code, from **request to response**, still works for simple apps, but using a modern architecture has many benefits.
+Enter the **middleware**, an intermediary code that processes the request in various ways before it even reaches your custom code.
+
+<!--- EXTENDED -->
+
+In this article we will list the advantages of using middlewares in your applications.
+From maintainability, to future-proof design, the benefits can't be ignored.
+
+### What does middleware do?
+
+The versatility of middleware is virtually endless.
+
+- Do you need it to pre-process your request by validating the user session or the form input?
+- Do you need to cache or translate the request?
+- Do you need to contact a third party site to enhance the request?
+
+Your custom middleware can handle **any specific logic** your application may require.
+Some useful things to implement are error handling, CORS, routing, translating.
+
+Ultimately, you can split your application into multiple middlewares that process the request before it reaches your main code.
+An added benefit is that the middlewares can be **reused more easily** in other applications.
+Even better, the middlewares **don't need to be written in the same programming language** or coding architecture.
```

Comment:
This one is ... not quite right.

There are two types of middleware architectures:

- One used within a framework to act as a transformative pipeline to transform a request to a response (that's where Mezzio sits)
- One that acts as a filter/translation/router for data, ultimately determining if data needs to be handled, and, if so, where to funnel it (this is the realm of Kafka)

I'd likely remove this sentence, or provide these clarifications.

Reply:
@weierophinney , i hope it's ok to add your comment in there. please let me know if you want me to edit anything